src/loaders/RawArrayLoader.ts

```diff
@@ -0,0 +1,143 @@
+  physicalPixelSize: new Vector3(json.physicalPixelSize[0], json.physicalPixelSize[1], json.physicalPixelSize[2]),
+  spatialUnit: json.spatialUnit || "μm",
+
+  numChannels: json.sizeC,
+  channelNames: json.channelNames,
+  channelColors: undefined, //json.channelColors,
+
+  times: 1,
+  timeScale: 1,
+  timeUnit: "s",
+
+  numMultiscaleLevels: 1,
+  multiscaleLevel: 0,
+
+  transform: {
+    translation: new Vector3(0, 0, 0),
+    rotation: new Vector3(0, 0, 0),
+  },
+
+  userData: json.userData,
+});
+
+class RawArrayLoader implements IVolumeLoader {
+  // one array per channel
+  data: RawArrayData;
+  jsonInfo: RawArrayInfo;
+
+  cache?: VolumeCache;
+
+  constructor(rawData: RawArrayData, rawDataInfo: RawArrayInfo, cache?: VolumeCache) {
+    this.jsonInfo = rawDataInfo;
+    this.data = rawData;
+    // check consistent dims
+    if (
+      this.data.shape[0] !== this.jsonInfo.sizeC ||
+      this.data.shape[1] !== this.jsonInfo.sizeZ ||
+      this.data.shape[2] !== this.jsonInfo.sizeY ||
+      this.data.shape[3] !== this.jsonInfo.sizeX
+    ) {
+      throw new Error("RawArrayLoader: data shape does not match metadata");
+    }
+    this.cache = cache;
+  }
+
+  async loadDims(_loadSpec: LoadSpec): Promise<VolumeDims[]> {
+    const jsonInfo = this.jsonInfo;
+
+    const d = new VolumeDims();
+    d.shape = [1, jsonInfo.sizeC, jsonInfo.sizeZ, jsonInfo.sizeY, jsonInfo.sizeX];
+    d.spacing = [1, 1, jsonInfo.physicalPixelSize[2], jsonInfo.physicalPixelSize[1], jsonInfo.physicalPixelSize[0]];
+    d.spaceUnit = jsonInfo.spatialUnit || "μm";
+    d.dataType = "uint8";
+    return [d];
+  }
+
+  async createVolume(loadSpec: LoadSpec, onChannelLoaded?: PerChannelCallback): Promise<Volume> {
+    const jsonInfo = this.jsonInfo;
+    const imageInfo = convertImageInfo(jsonInfo);
+
+    const vol = new Volume(imageInfo, loadSpec, this);
+    vol.channelLoadCallback = onChannelLoaded;
+    vol.imageMetadata = buildDefaultMetadata(imageInfo);
+    return vol;
+  }
+
+  async loadVolumeData(vol: Volume, explicitLoadSpec?: LoadSpec, onChannelLoaded?: PerChannelCallback): Promise<void> {
+    const loadSpec = explicitLoadSpec || vol.loadSpec;
+
+    const requestedChannels = loadSpec.channels;
+
+    vol.loadSpec = {
+      ...loadSpec,
+      // `subregion` and `multiscaleLevel` are unused by this loader
+      subregion: new Box3(new Vector3(0, 0, 0), new Vector3(1, 1, 1)),
+      multiscaleLevel: 0,
+      // TODO force include ALL channels always?
+      //channels: requestedChannels,
+    };
+
+    const imageInfo = vol.imageInfo;
+    const cache = this.cache;
+    for (let chindex = 0; chindex < imageInfo.numChannels; ++chindex) {
+      if (requestedChannels && requestedChannels.length > 0 && !requestedChannels.includes(chindex)) {
+        continue;
+      }
+      // TODO this loader is its own cache and need not even use the cache?
+      const cacheResult = cache?.get(`${imageInfo.name}/${chindex}`);
```

Comment:
It seems like this would cause the memory footprint of this loader to double if it was loading from a URL resource? Or would it just ignore it if the provided URL isn't a real URL...?

Reply:
Agree with the TODO that this loader probably doesn't need access to the cache. Though I could see wanting to let the cache count it towards its maximum size in order to keep memory use under control? In that case, though, I'd just insert the whole contents of `data` as one entry - it doesn't seem all that worth it to avoid the work of splitting the array into different channels.